src/main/java/org/apache/datasketches/filters/bloomfilter/BitArray.java

```diff
@@ -17,194 +17,111 @@
 
-  // may need to recompute value:
-  // O(1) if only getAndSetBit() has been used
-  // O(data_.length) if setBit() has ever been used
-  long getNumBitsSet() {
-    if (isDirty_) {
-      numBitsSet_ = 0;
-      for (final long val : data_) {
-        numBitsSet_ += Long.bitCount(val);
-      }
-    }
-    return numBitsSet_;
-  }
-
-  long getCapacity() { return (long) data_.length * Long.SIZE; }
+  abstract boolean getBit(final long index);
 
-  int getArrayLength() { return data_.length; }
+  abstract boolean getAndSetBit(final long index);
 
-  // applies logical OR
-  void union(final BitArray other) {
-    if (data_.length != other.data_.length) {
-      throw new SketchesArgumentException("Cannot union bit arrays with unequal lengths");
-    }
+  abstract void setBit(final long index);
 
-    numBitsSet_ = 0;
-    for (int i = 0; i < data_.length; ++i) {
-      data_[i] |= other.data_[i];
-      numBitsSet_ += Long.bitCount(data_[i]);
-    }
-    isDirty_ = false;
-  }
+  abstract long getNumBitsSet();
 
-  // applies logical AND
-  void intersect(final BitArray other) {
-    if (data_.length != other.data_.length) {
-      throw new SketchesArgumentException("Cannot intersect bit arrays with unequal lengths");
-    }
+  abstract void reset();
 
-    numBitsSet_ = 0;
-    for (int i = 0; i < data_.length; ++i) {
-      data_[i] &= other.data_[i];
-      numBitsSet_ += Long.bitCount(data_[i]);
-    }
-    isDirty_ = false;
-  }
+  abstract long getCapacity();
 
-  // applies bitwise inversion
-  void invert() {
-    if (isDirty_) {
-      numBitsSet_ = 0;
-      for (int i = 0; i < data_.length; ++i) {
-        data_[i] = ~data_[i];
-        numBitsSet_ += Long.bitCount(data_[i]);
-      }
-      isDirty_ = false;
-    } else {
-      for (int i = 0; i < data_.length; ++i) {
-        data_[i] = ~data_[i];
-      }
-      numBitsSet_ = getCapacity() - numBitsSet_;
-    }
-  }
+  abstract int getArrayLength();
 
-  long getSerializedSizeBytes() {
-    // We only really need an int for array length but this will keep everything
-    // aligned to 8 bytes.
-    // Always write array length and numBitsSet, even if empty
-    return isEmpty() ? Long.BYTES : Long.BYTES * (2L + data_.length);
-  }
+  abstract void union(final BitArray other);
 
-  void writeToBuffer(final WritableBuffer wbuf) {
-    wbuf.putInt(data_.length);
-    wbuf.putInt(0); // unused
+  abstract void intersect(final BitArray other);
 
-    if (!isEmpty()) {
-      wbuf.putLong(isDirty_ ? -1 : numBitsSet_);
-      wbuf.putLongArray(data_, 0, data_.length);
-    }
-  }
+  abstract void invert();
 
   // prints the raw BitArray as 0s and 1s, one long per row
   @Override
   public String toString() {
     final StringBuilder sb = new StringBuilder();
-    for (int i = 0; i < data_.length; ++i) {
+    for (int i = 0; i < getArrayLength(); ++i) {
       sb.append(i + ": ")
-        .append(printLong(data_[i]))
+        .append(printLong(getLong(i)))
         .append("\n");
```

Comment:
Just as a comment: we should always use the Java "line.separator" instead of "\n" for new lines, because the Java one is automatically adjusted based on the platform.  "\n" works on unix and MacOS, but Windows requires "\r\n".
I am not perfect in this either, but try to keep it in mind. You can acquire the Java "line.separator" like this:
`public static final String LS = System.getProperty("line.separator");`

Reply:
Thanks for catching this. I forgot to update `toString()` properly after my initial uses for debugging in other ways, too, but missed this one on a subsequent pass, too.